Draft: FIM permutations

Stillerman opened the pull request:
Here is a draft that adds optional FIM permutations while finetuning. Some things I was hoping to get feedback on

- I added the permuting as part of the `ConstantLengthDataset` class. Is that the right place to do this?

- With FIM enabled, batches take ~2x as long to load. Is this acceptable? Here are the stats for loading 5 batches with and without FIM on my Intel MacBook Pro
```
FIM time: 66.68208975399999
Non-FIM time: 32.142377584
```

- I cannot actually test that the training still works on the script because I don't have access to a GPU. The training seems to work in the [Colab I adapted](https://colab.research.google.com/drive/1CADU0YxD3p7pLivgST0toYFLUdGWDP5f) but this should be checked to make sure all is still well

arjunguha commented:
I've read this and it seems okay. But, someone should actually run it on a GPU.

Stillerman commented:
Switched from character-based FIM to token-based. Now we do not have to do expensive encode/decode and it is not slowed down at all.

```
FIM time: 37.271945165999995
Non-FIM time: 37.641369552
```

Will test on GPU soon

Stillerman commented:
Finetuned for 48 hours on an A100 on the ruby stack and got [this](https://huggingface.co/stillerman/santacoder-ruby).

MultiPL-E pass@1 was 0.1

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
 
     return total_characters / total_tokens
 
+# this is expensive so we cache it
```
I suggest moving the `get_fim_token_ids` and `permute` functions to a new file `fim_utils.py` to keep this file short and clear.

Stillerman replied:
done

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
+    Maintain the same sample length (if transform creates a few extra tokens, drop them).
+    """
+
+    suffix_tok_id, prefix_tok_id, middle_tok_id, pad_tok_id, = get_fim_token_ids(tokenizer)
```
we can call this function once in the dataset class and pass values as args to `permute` to avoid calling it for each sample

Stillerman replied:
done

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
+        try:
+            # A boundary can be =0 (prefix will be empty)
+            # a boundary can be =len(contents) (suffix will be empty)
+            # The two boundaries can be equal (middle will be empty)
```
```suggestion
    if np_rng.binomial(1, fim_rate):
        try:
```

Stillerman replied:
👍

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
+            # The two boundaries can be equal (middle will be empty)
+            boundaries = list(np_rng.randint(low=0, high=len(sample) + 1, size=2))
+            boundaries.sort()
+        except ValueError as e:
```
does this ever happen in the training?

Stillerman replied:
Not that I saw, and I can't imagine how it would. Removed.

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
+        middle = np.array(sample[boundaries[0]:boundaries[1]], dtype=np.int64)
+        suffix = np.array(sample[boundaries[1]:], dtype=np.int64)
+
+        # here we truncate each given segment to fit the same length as it was before
+        # A consequence is that we never reach the end of a file?
+        # we should rather truncate at the context-level
```
we can probably remove these comments if we're not implementing it
```suggestion
```

Stillerman replied:
👍

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
+            new_length = suffix.shape[0] + prefix.shape[0] + middle.shape[0] + 3
+            diff = new_length - len(sample)
+            if diff > 0: # too long
+                if suffix.shape[0] <= diff: # if there's no space to truncate the suffix: stop and report it. atm i should have stopped this from happening
```
same comment as above
```suggestion
                if suffix.shape[0] <= diff:
```

Stillerman replied:
👍

loubnabnl commented on train.py:
```diff
@@ -119,7 +194,13 @@ def __iter__(self):
             tokenized_inputs = self.tokenizer(buffer, truncation=False)["input_ids"]
             all_token_ids = []
+
+            np_rng = np.random.RandomState(seed=555)
```
let's use the `seed` provided in the arguments and pass it as input to the class

Stillerman replied:
👍

loubnabnl commented on train.py:
```diff
@@ -169,8 +252,21 @@ def create_datasets(tokenizer, args):
     )
-    return train_dataset, valid_dataset
+
+    valid_dataset_no_fim = ConstantLengthDataset(
```
Did you want to add an option to compute validation loss in non FIM mode? Currently this dataset is not used in the script

loubnabnl commented on train.py:
```diff
@@ -219,7 +315,14 @@ def run_training(args, train_data, val_data):
 def main(args):
     tokenizer = AutoTokenizer.from_pretrained(args.model_path, use_auth_token=True)
-    train_dataset, eval_dataset = create_datasets(tokenizer, args)
+    train_dataset, eval_dataset, eval_no_fim = create_datasets(tokenizer, args)
```
`eval_no_fim` is not used (see comment above). We can add an option if the user wants to evaluate on both datasets when `fim_rate` is not zero. To do that, you can pass a `dict` as `eval_dataset` to the `Trainer` ([see](eval_no_fim))

Stillerman replied:
That was left over from testing. I don't see why we would want non-FIM eval so I have removed for now. Is this something we would want?

loubnabnl commented on train.py:
```diff
@@ -219,7 +315,14 @@ def run_training(args, train_data, val_data):
-    train_dataset, eval_dataset = create_datasets(tokenizer, args)
+    train_dataset, eval_dataset, eval_no_fim = create_datasets(tokenizer, args)
+
+    fim_time = timeit.timeit(lambda: next(iter(eval_dataset)), number=5)
+    non_fim_time = timeit.timeit(lambda: next(iter(eval_no_fim)), number=5)
+
+    print(f"FIM time: {fim_time}")
+    print(f"Non-FIM time: {non_fim_time}")
+
```
You didn't observe a slow down in CPU nor GPU, right? If so we can remove this
```suggestion
```

Stillerman replied:
Removed

loubnabnl commented on train.py:
```diff
@@ -66,6 +72,71 @@ def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
+## Adapted from https://github.com/bigcode-project/Megatron-LM/blob/6c4bf908df8fd86b4977f54bf5b8bd4b521003d1/megatron/data/gpt_dataset.py
+def permute(sample, np_rng, tokenizer, fim_rate=0.5, fim_spm_rate=0.5, truncate_or_pad=False):
+    """
+    Take in a sample (np array w/ size (0,chunklength)) and perform a FIM transformation on it. 
```
```suggestion
    Take in a sample (list of tokens) and perform a FIM transformation on it with a probability of fim_rate, using two FIM modes: PSM and SPM (with a probability of fim_spm_rate). 
```

Stillerman replied:
👍

loubnabnl reviewed (commented):
Thanks a lot for this contribution! The code looks good, I just left some small comments mainly about the format. 
Let's move the FIM functions to a separate file to keep the main file concise and easy. Can you also run `isort` and `black` on the file?

Stillerman commented:
I addressed all of the comments in that last commit. The only change really of note is that I removed the `eval_no_fim` instead of making that an additional eval set, as that was left over from testing. Is there some reason why we would want to add it as an additional eval set?

Also, reran a toy finetune for an hour and everything seems to be working same as it did before.

Stillerman commented:
Finetuned Julia for 2m tokens both with and without FIM and tested both with MultiPL-E.

[With FIM](https://huggingface.co/stillerman/santacoder-julia-fim) pass@1 = 0.1

[Without FIM](https://huggingface.co/stillerman/santacoder-julia-no-fim) pass@1 = 0.09

loubnabnl commented:
Looks great! Thanks for working on this and for your patience!
